File: Data Structures & Algorithms/interleaving-string/submission-18.java
class Solution {
    
    public boolean isInterleave(String s1, String s2, String s3) {
        int M = s1.length(), N = s2.length();
        boolean[] dp = new boolean[N+1];
        if(M+N != s3.length())
            return false;
        
        for(int i=M;i>=0;i--){
            boolean[] row = new boolean[N+1];
            if(i==M)
                row[N] = true;
            for(int j=N;j>=0;j--){
                if(row[j])
                    continue;
                int k = i+j;
                row[j] = (i<M && s1.charAt(i)==s3.charAt(k)?dp[j]:false) || 
                    (j<N && s2.charAt(j)==s3.charAt(k)?row[j+1]:false);
            }
            dp = row;
        }
        return dp[0];
        

        
    }

    // public boolean dfs(int i,int j,String s1,String s2,String s3){
    //     if(i==s1.length() && j == s2.length()){
    //         return true;
    //     }
    //     int k = i+j;
    //     if(dp[i][j]!=null)
    //         return dp[i][j];
    //     if(i<s1.length() && s1.charAt(i)==s3.charAt(k) && dfs(i+1,j,s1,s2,s3)){
    //         dp[i][j] = true;
    //         return true;
    //     } 
    //     if(j<s2.length() && s2.charAt(j)==s3.charAt(k) && dfs(i,j+1,s1,s2,s3)){
    //         dp[i][j] = true;
    //         return true;
    //     }
    //     dp[i][j] = false;
    //     return false;
    // }
}

File: Data Structures & Algorithms/count-subsequences/submission-18.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        M,N = len(s),len(t)
        if N>M:
            return 0
        dp = [0] * (N+1)
        dp[N] = 1

        for i in range(M-1,-1,-1):
            cur = [0] * (N+1)
            cur[N] = 1
            for j in range(N-1,-1,-1):
                cur[j] = dp[j]
                if s[i] == t[j]:
                    cur[j] += dp[j+1]
            dp = cur
        return dp[0]
        # def dfs(i,j):
        #     if j == N:
        #         return 1
        #     if i == M:
        #         return 0
        #     if (i,j) in dp:
        #         return dp[(i,j)]
        #     res = dfs(i+1,j)
        #     if s[i] == t[j]:
        #         res += dfs(i+1,j+1)
        #     dp[(i,j)] = res
        #     return dp[(i,j)]
        # return dfs(0,0)




      




        